### PR TITLE
feat: jumpstart EULA models

### PR DESCRIPTION
**Pull request by evakravi**
*Issue #, if available:*

*Description of changes:*
JumpStart is releasing models that require the customer to accept an end-user license agreement (EULA). These models have protected model weights, so the artifacts cannot be stored in public S3 buckets. 

Models that require EULA will rely use a `ModelPackage` to store the model weights. This paradigm requires modifying the metadata for JumpStart models, as well as modifying the workflow so that if a model package arn is detected, a Model is created from the model package versus the normal method of injecting public s3 artifacts into the container. 

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [ ] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [ ] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/jumpstart/artifacts/model_uris.py [hunk 1]
```diff
@@ -76,18 +94,35 @@ def _retrieve_model_uri(
...
+
     if model_scope == JumpStartScriptScope.INFERENCE:
-        model_artifact_key = (
-            getattr(model_specs, "hosting_prepacked_artifact_key", None)
-            or model_specs.hosting_artifact_key
-        )
+
+        model_artifact_key = None
+
+        for model_artifact_field_name in _MODEL_ARTIFACT_PRIORITY_LIST:
```
Comment (akozd) [inline src/sagemaker/jumpstart/artifacts/model_uris.py:106]: just from a code legibility standpoint this block would benefit from being extracted into a helper function

File: src/sagemaker/jumpstart/artifacts/script_uris.py [hunk 2]
```diff
@@ -76,16 +80,27 @@ def _retrieve_script_uri(
...
     )
 
+    bucket_override = os.environ.get(ENV_VARIABLE_JUMPSTART_SCRIPT_ARTIFACT_BUCKET_OVERRIDE)
+
+    if bucket_override:
+        logger.info("Using JumpStart bucket override for script URI: '%s'", bucket_override)
+
     if script_scope == JumpStartScriptScope.INFERENCE:
+        script_uri = getattr(model_specs, "hosting_script_uri", None)
+        if script_uri is not None:
```
Comment (akozd) [inline src/sagemaker/jumpstart/artifacts/script_uris.py:90]: Same here, I'm not a fan of the nested ifs. Not a blocker through

**Review by akozd — APPROVED** [review]

File: src/sagemaker/jumpstart/utils.py [hunk 3]
```diff
@@ -402,6 +402,12 @@ def verify_model_region_and_return_specs(
             f"JumpStart model ID '{model_id}' and version '{version}' " "does not support training."
         )
 
+    if model_specs.eula_model:
+        LOGGER.info(
```
Comment (whittech1) [inline src/sagemaker/jumpstart/utils.py:406]: See prior comment, we can perhaps give them the ability to retrieve the EULA from our public bucket

**Review by whittech1 — APPROVED** [review]

File: src/sagemaker/jumpstart/utils.py [hunk 4]
```diff
@@ -402,6 +402,12 @@ def verify_model_region_and_return_specs(
...
+            "Using model with end-user license agreement (EULA). "
+            "Deploying this model requires accepting EULA terms."
```
Comment (whittech1) [inline src/sagemaker/jumpstart/utils.py:408]: May want to provide a more exact call to action

**Review by whittech1 — APPROVED** [review]

**Review by whittech1 — APPROVED** [review]

**Comment by akrishna1995** [comment]
@evakravi Please remove from draft and ping me when you need this to be reviewed from sagemaker team - Thanks

File: tests/unit/sagemaker/jumpstart/constants.py [hunk 5]
```diff
@@ -2299,6 +2299,8 @@
     "training_script_key": "source-directory-tarballs/pytorch/transfer_learning/ic/v1.0.0/sourcedir.tar.gz",
     "training_prepacked_script_key": None,
     "hosting_prepacked_artifact_key": None,
+    "hosting_model_package_arns": None,
+    "hosting_eula_key": False,
```
Comment (knikure) [inline tests/unit/sagemaker/jumpstart/constants.py:2303]: We will have to change it to None

**Review by knikure — COMMENTED** [review]
Changes looks good to me. Once this change is public we are planning to have a follow-up PR to add integ-test as well as unit tests under jumpstart module.

**Review by knikure — APPROVED** [review]

**Review by Captainia — APPROVED** [review]

**Review by akozd — APPROVED** [review]

File: src/sagemaker/jumpstart/model.py [hunk 6]
```diff
@@ -301,9 +307,49 @@ def _is_valid_model_id_hook():
         self.tolerate_vulnerable_model = model_init_kwargs.tolerate_vulnerable_model
         self.tolerate_deprecated_model = model_init_kwargs.tolerate_deprecated_model
         self.region = model_init_kwargs.region
+        self.model_package_arn = model_init_kwargs.model_package_arn
 
         super(JumpStartModel, self).__init__(**model_init_kwargs.to_kwargs_dict())
 
+    def _create_sagemaker_model(self, *args, **kwargs):  # pylint: disable=unused-argument
+        """Create a SageMaker Model Entity
```
Comment (JGuinegagne) [inline src/sagemaker/jumpstart/model.py:315]: nit^2: missing period `.` at the end of the docstring.

File: src/sagemaker/jumpstart/utils.py [hunk 7]
```diff
@@ -402,6 +402,17 @@ def verify_model_region_and_return_specs(
...
         )
 
+    if model_specs.hosting_eula_key and scope == constants.JumpStartScriptScope.INFERENCE.value:
+        LOGGER.info(
+            "Model '%s' requires accepting end-user license agreement (EULA). "
+            "See https://%s.s3.%s.amazonaws.com%s/%s for terms of use.",
+            model_id,
+            get_jumpstart_content_bucket(region=region),
+            region,
+            ".cn" if region.startswith("cn-") else "",
```
Comment (JGuinegagne) [inline src/sagemaker/jumpstart/utils.py:412]: callout: this may not work for gov-cloud.

**Review by JGuinegagne — COMMENTED** [review]
Could you also add unit tests for model package flow as well?

**Pull request merged**